src/algorithms.jl

```diff
@@ -167,13 +171,51 @@ the final `f!(A, out, alg::Algorithm)`.
     f!(A, alg::Algorithm)
 ```
 
+The number of inputs can be set with the `n_args` keyword
+argument, so that 
+
+```julia
+@functiondef n_args=2 f
+```
+
+would create 
+
+```julia
+    f(A, B; kwargs...)
+    f(A, B, alg::Algorithm)
+    f!(A, B, [out]; kwargs...)
+    f!(A, B, alg::Algorithm)
+```
+
 See also [`copy_input`](@ref), [`select_algorithm`](@ref) and [`initialize_output`](@ref).
 """
-macro functiondef(f)
+macro functiondef(args...)
+    kwargs = map(args[1:end-1]) do kwarg
+        if kwarg isa Symbol
+            :($kwarg = $kwarg)
+        elseif Meta.isexpr(kwarg, :(=))
+            kwarg
+        else
+            throw(ArgumentError("Invalid keyword argument '$kwarg'"))
+        end
+    end
+    isempty(kwargs) || length(kwargs) == 1 || throw(ArgumentError("Only one keyword argument to `@functiondef` is supported"))
+    f_n_args = 1 # default
+    if length(kwargs) == 1
+        kwarg = only(kwargs) # only one kwarg is currently supported, TODO modify if we support more
+        key::Symbol, val = kwarg.args
+        key === :n_args || throw(ArgumentError("Unsupported keyword argument $key to `@functiondef`"))
+        (isa(val, Integer) && val > 0) || throw(ArgumentError("`n_args` keyword argument to `@functiondef` should be an integer > 0"))
+        f_n_args = val
+    end
+
+    f = args[end]
     f isa Symbol || throw(ArgumentError("Unsupported usage of `@functiondef`"))
     f! = Symbol(f, :!)
 
-    ex = quote
+    # TODO is the right way?
+    @gensym A B
```

Comment:
I'm not sure if this is necessary, I think the scoping rules always make sure that the `A` and `B` refer to the function arguments. (Also I don't think you are using these right now, since it would have to be interpolated in the quoted expression)

Reply:
Ah yes whoops 😅